### ISSUE TITLE
Parameters and Hyperparameters

### ISSUE THREAD

**Issue by huseingharibshah** [huseingharibshah/Parameters#1]
✅ هایپرپارامترها در یادگیری ماشین چیستند؟

در مدل‌های یادگیری ماشین دو نوع مقدار وجود دارد: پارامترها (Parameters) و هایپرپارامترها (Hyperparameters).

🔹 هایپرپارامترها (Hyperparameters)

مقدارهایی هستند که قبل از شروع آموزش مدل مشخص می‌شوند و در طول آموزش توسط مدل یاد گرفته نمی‌شوند.
این مقادیر رفتار و ساختار مدل را تعیین می‌کنند.

مثال‌های معروف از هایپرپارامترها:

Learning Rate در شبکه‌های عصبی

Number of Trees در Random Forest

Max Depth در درخت تصمیم

Batch Size در Deep Learning

Number of Neighbors (k) در KNN

C و γ در SVM

Number of Epochs

مدل این مقادیر را خودش یاد نمی‌گیرد؛ ما باید تعیین کنیم.

---

🔵 پارامتر چیست؟ (Parameters)

پارامترها مقادیری هستند که مدل در هنگام آموزش آن‌ها را یاد می‌گیرد تا بهترین عملکرد را داشته باشد.

مثال‌های پارامتر:

وزن‌ها (Weights) در شبکه عصبی

ضرایب (Coefficients) در رگرسیون خطی

احتمال‌ها در مدل Naive Bayes

مقادیر تقسیم‌کننده در درخت تصمیم

پارامترها بعد از آموزش مدل در خروجی ذخیره می‌شوند و مدل با استفاده از آن‌ها پیش‌بینی می‌کند.

---

🎯 فرق بین پارامتر و هایپرپارامتر

ویژگی	پارامتر	هایپرپارامتر

چه کسی آن را تعیین می‌کند؟	مدل یاد می‌گیرد	انسان تعیین می‌کند
کی تعیین می‌شود؟	هنگام آموزش	قبل از آموزش
مربوط به چه چیزی است؟	درونیات مدل	ساختار و رفتار مدل
مثال	وزن‌ها، ضرایب	learning rate، max depth، تعداد epoch

---

🌟 چگونه مقدار هایپرپارامترها انتخاب می‌شود؟

برای انتخاب مقدارهای خوب، معمولاً از روش‌های جستجو استفاده می‌کنیم:

1️⃣ Grid Search

همه ترکیب‌های ممکن از مجموعه‌ای از مقدارها را امتحان می‌کند.

دقیق اما کند.

مثال:
learning_rate = {0.01, 0.1, 1}
max_depth = {3, 5}

---

2️⃣ Random Search

به جای تست همه حالت‌ها، به صورت تصادفی چند حالت را امتحان می‌کند.

عملکرد تقریباً خوب + سریع‌تر از Grid.

---

3️⃣ Bayesian Optimization

هوشمندانه مقادیر را انتخاب می‌کند و از نتایج قبلی استفاده می‌کند.

مناسب پروژه‌های واقعی و مدل‌های سنگین.

---

4️⃣ اختصاصی برای DL → روش‌های تنظیم یادگیری

Learning Rate Schedule

Adaptive Optimizers (Adam, RMSProp)

Early Stopping
